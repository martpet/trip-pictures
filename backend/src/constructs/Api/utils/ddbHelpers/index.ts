export * from './ddbBatchGet';
export * from './ddbBatchWrite';
export * from './ddbScan';
