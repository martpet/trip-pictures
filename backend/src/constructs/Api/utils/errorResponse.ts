import { APIGatewayProxyResult } from 'aws-lambda';
import { getReasonPhrase, StatusCodes } from 'http-status-codes';
import { SetRequired } from 'type-fest';

import { ApiErrorResponseBody } from '~/types';

type BaseOptions = {
  description?: string;
  error?: unknown;
  statusCode?: StatusCodes;
};

type WithExposedError = SetRequired<BaseOptions, 'error'> & {
  exposeError: true;
};

type WithoutExposedError = BaseOptions & {
  exposeError?: never | false;
};

type ErrorResponseOptions = WithExposedError | WithoutExposedError;

export const errorResponse = (
  traceId: string,
  options: ErrorResponseOptions = {}
): APIGatewayProxyResult => {
  const {
    statusCode = StatusCodes.INTERNAL_SERVER_ERROR,
    description,
    error,
    exposeError,
  } = options;

  const { envName } = globalLambdaProps;

  const errObj: ApiErrorResponseBody['error'] = {
    statusCode,
    message: getReasonPhrase(statusCode),
    traceId,
  };

  if (description) {
    errObj.description = description;
  }

  if (error && (exposeError ?? envName !== 'production')) {
    errObj.error = String(error);
  }

  const body: ApiErrorResponseBody = {
    error: errObj,
  };

  console.error({ traceId, error });

  return {
    statusCode,
    body: JSON.stringify(body),
  };
};
