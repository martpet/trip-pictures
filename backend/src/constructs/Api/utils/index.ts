export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
