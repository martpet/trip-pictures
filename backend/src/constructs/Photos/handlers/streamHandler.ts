import {
  createPhotoS3Key,
  DynamoDBRecord,
  DynamoDBStreamHandler,
  Merge,
  PhotosTableItem,
  PutObjectTaggingCommand,
  S3Client,
} from 'lambda-layer';

import {
  IS_PHOTOS_TABLE_ITEM_CREATED_S3_TAG,
  STREAM_HANDLER_PHOTO_BUCKET_ENV_KEY,
} from '../consts';

const s3Client = new S3Client({});

type PhotoRecord = Merge<
  DynamoDBRecord,
  { dynamodb: { NewImage: Record<keyof PhotosTableItem, { S: string }> } }
>;

export const handler: DynamoDBStreamHandler = async (event) => {
  const photoBucket = process.env[STREAM_HANDLER_PHOTO_BUCKET_ENV_KEY];

  const photosRecords = event.Records.filter(
    ({ dynamodb }) => dynamodb?.NewImage
  ) as PhotoRecord[];

  console.info('Pending items:', photosRecords.length);

  const promises = photosRecords.map(async ({ dynamodb }) => {
    const s3Key = createPhotoS3Key({
      fingerprint: dynamodb.NewImage.fingerprint.S,
      userId: dynamodb.NewImage.userId.S,
    });

    return s3Client.send(
      new PutObjectTaggingCommand({
        Bucket: photoBucket,
        Key: s3Key,
        Tagging: {
          TagSet: [{ Key: IS_PHOTOS_TABLE_ITEM_CREATED_S3_TAG, Value: 'true' }],
        },
      })
    );
  });

  await Promise.all(promises);
};
