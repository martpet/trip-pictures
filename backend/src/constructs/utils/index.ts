export * from './createEdgeFunction';
export * from './createNodejsFunction';
export * from './createTable';
