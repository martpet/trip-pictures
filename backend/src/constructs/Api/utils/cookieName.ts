import { AllCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllCookies) => name;
