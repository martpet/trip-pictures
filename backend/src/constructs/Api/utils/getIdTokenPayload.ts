import { IdTokenPayload } from '~/constructs/Api/types';

export const getIdTokenPayload = async (idToken: string): Promise<IdTokenPayload> => {
  const camelcaseKeys = (await import('camelcase-keys')).default;
  const payload = JSON.parse(Buffer.from(idToken.split('.')[1], 'base64').toString());

  return camelcaseKeys(payload);
};
