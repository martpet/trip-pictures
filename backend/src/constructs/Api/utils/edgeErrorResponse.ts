import { getReasonPhrase } from 'http-status-codes';

import { LambdaEdgeResponse } from '~/constructs/Api/types';
import { errorResponse } from '~/constructs/Api/utils/errorResponse';

export const edgeErrorResponse = (
  ...errorResponseArgs: Parameters<typeof errorResponse>
): LambdaEdgeResponse => {
  const { statusCode, body } = errorResponse(...errorResponseArgs);

  const response: LambdaEdgeResponse = {
    status: statusCode,
    statusDescription: getReasonPhrase(statusCode),
    headers: { 'content-type': [{ value: 'application/json' }] },
    body,
  };

  return response;
};
