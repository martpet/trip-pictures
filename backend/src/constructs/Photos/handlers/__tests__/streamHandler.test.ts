import { PutObjectTaggingCommand, S3Client } from '@aws-sdk/client-s3';
import { mockClient } from 'aws-sdk-client-mock';

import {
  itCalls,
  itRejects,
  itResolves,
  itSendsAwsCommand,
} from '~/constructs/Api/utils';
import { createPhotoS3Key } from '~/utils';

import { handler } from '../streamHandler';

vi.mock('~/utils/createPhotoS3Key');

const s3Mock = mockClient(S3Client);

process.env.photoBucket = 'dummyPhotoBucketName';

beforeEach(() => {
  s3Mock.reset();

  s3Mock.on(PutObjectTaggingCommand).resolves({});
});

const args = [
  {
    Records: [
      {
        dynamodb: {
          NewImage: {
            fingerprint: { S: 'dummyFingerPrint1' },
            userId: { S: 'dummyUserId1' },
          },
        },
      },
      {
        dynamodb: {
          NewImage: {
            fingerprint: { S: 'dummyFingerPrint2' },
            userId: { S: 'dummyUserId2' },
          },
        },
      },
    ],
  },
] as unknown as Parameters<typeof handler>;

describe('streamHandler', () => {
  itCalls(createPhotoS3Key, handler, args);
  itSendsAwsCommand(PutObjectTaggingCommand, s3Mock, handler, args);
  itResolves(handler, args);

  describe('when PutObjectTaggingCommand returns a rejected promise', () => {
    beforeEach(() => {
      s3Mock
        .on(PutObjectTaggingCommand)
        .rejectsOnce('dummy PutObjectTaggingCommand error');
    });
    itRejects(handler, args);
  });
});
