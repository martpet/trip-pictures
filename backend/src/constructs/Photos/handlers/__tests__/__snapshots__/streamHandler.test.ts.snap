// Vitest Snapshot v1

exports[`streamHandler > calls "createPhotoS3Key" with correct args 1`] = `
[
  [
    {
      "fingerprint": "dummyFingerPrint1",
      "userId": "dummyUserId1",
    },
  ],
  [
    {
      "fingerprint": "dummyFingerPrint2",
      "userId": "dummyUserId2",
    },
  ],
]
`;

exports[`streamHandler > resolves with a correct value 1`] = `undefined`;

exports[`streamHandler > sends "PutObjectTaggingCommand" from "S3Client" with correct args 1`] = `
{
  "Bucket": "dummyPhotoBucketName",
  "Key": "dummyPhotoS3Key",
  "Tagging": {
    "TagSet": [
      {
        "Key": "isTableItemCreated",
        "Value": "true",
      },
    ],
  },
}
`;

exports[`streamHandler > when PutObjectTaggingCommand returns a rejected promise > rejects with a correct value 1`] = `[Error: dummy PutObjectTaggingCommand error]`;
